### ISSUE TITLE
Breakpoint checkboxes have wrong role

### ISSUE THREAD

**Issue by AccessibilityTestingTeam-TCS** [microsoft/vscode#52390]
**Environment Details:**
VSCode Version : 1.24.0
OS Version : HighSierra10.13.5

**Additional Details:**
MAS Violated: MAS 1.3.1
ScreenReader : VoiceOver

**Repro Steps:**
1. Launch VS Code while VoiceOver is ON.
2. Navigate to Activity Bar and select "View Debug " button.
3. Navigate to "All Exceptions" and "User Handled Exceptions" checkboxes in "Breakpoints" section.

**Actual:**
Screenreader reads the tick boxes as list boxes. The role of the "All Exceptions" and "User Handled Exceptions" controls is incorrect.

**Expected:**
Screenreader should read the proper role of the controls as tickboxes so that proper action can be performed by the user.

**Recommendations:**
Provide proper role to the "All Exceptions" and "User Handled Exceptions" controls as tick boxes.(checkboxes)
or,
Refer below link which is repository of bug fixes code snippets:
https://microsoft.sharepoint.com/teams/msenable/mas/pages/browse-fixes.aspx

**User Impact:**
The screenreader users will not be able to know the correct role of the controls to perform appropriate actions on the controls present on the screen.

**MAS Reference:**
MAS 1.3.1 - https://microsoft.sharepoint.com/teams/msenable/_layouts/15/WopiFrame.aspx?sourcedoc={54f28d1f-a2d1-4dcd-84e1-5c9b87e8aba4}

**Attachment For Reference:**
[A11y_VSCode_ViewDebug_VoiceOver_TickboxAsListBox.pptx](https://github.com/Microsoft/vscode/files/2118031/A11y_VSCode_ViewDebug_VoiceOver_TickboxAsListBox.pptx)

Does this issue occur when all extensions are disabled?: Yes

**Comment by isidorn** [comment]
For this to be done the list would have to support an option to pass in the aria role similar as to how it currently allow the aria-lebel to be passed.
Since from the breakpoints view I can not style the appropriate element.
@joaomoreno thoughts?

**Comment by joaomoreno** [comment]
Yeah we probably need that.

**Comment by isidorn** [comment]
@joaomoreno would you like a PR or you want to look into it?

**Comment by cleidigh** [comment]
I am doing a lot of CCing since various issues involve various people but they're all related:

@isidorn 
@joaomoreno 
@bpasero 
@roblourens 
I'm going to make an overall issue for this.  I have been doing a lot of playing around working on the setting side .  It has revealed several similar issues.  I also found a great resource that includes some important information that the official specs are slightly ambiguous.
http://whatsock.com/training/matrices/

Both list and tree widgets have several issues associated with this.  
- Both only use role=tree as the top level aria role .
- Widget rows are always role=treeitem
- Each tree item also has the attributes for level ,position and position in set to help the reader 
- My research and experimentation shows that editable/interactive elements are not clearly supported
  within a aria tree.  Both descendent element roles and their labeling become unpredictable despite
  some ambiguity in the specs.  I found it was implicit trees are always read-only/non- editable

as I said all do an issue for this, but here's my suggestions:

- Allow the tree widget to accept an option for the main role 
- Form based roles as top-level will not require any HTML structure change
- Use above for list also
- Investigate changing the main tree operation to use the newer treegrid role/structure
  This would require any HTML structure change using a table structure, this is explicitly editable
  and should be able to act like both or tree , list and even grid

Trying to coerce the current tree shows a bunch of inconsistencies even if certain things work sometimes
when the spec would imply they're not appropriate the reader still try to do their job.

**Comment by isidorn** [comment]
@cleidigh thanks for the reply, I have just discussed this with @joaomoreno and we have come up with a solution pretty similar to what you are prosposing - so we are on the same page.

1. We will add a getAriaRole to the [`IAccessibilityProvider`](https://github.com/Microsoft/vscode/blob/828d2179178549c744ea3d7f78e5086252713cf5/src/vs/base/parts/tree/browser/tree.ts#L435) of the Tree
2. We will add a IAccessibilityProvider to the List which can be passed as an option when the list is created. The IAccessiblityProvider will be able to pass the role the same way as for the tree

Those are the first steps. After we do that we can see what is also missing.
Anyways, we will not do this in endgame -> september

**Comment by cleidigh** [comment]
@isidorn 

Sounds good, some thoughts:

- IMHO the main widget role option should be an enum (tree, form, section, landmark) 
arbitrary roles could cause problems
- the "branch/row" roles should be derived from the main role above, again arbitrary could cause problems

If a main role does not support particular attributes, they should be removed within the widget automatically
e.g. in a form no levels, position et cetera

Is this something you are doing or would you like me to pick it up for September since I've worked on this already?

**Comment by roblourens** [comment]
@cleidigh do we need to get rid of the role now that we have a nicer way to do it? I tested with NVDA and it actually seems ok like this, not sure why.

**Comment by cleidigh** [comment]
@roblourens 
I'll take a look at this to see if any changes are necessary.

**Comment by cleidigh** [comment]
@roblourens 
I commented in the PR because I think were still missing the ability to set the role of the tree widget itself.
https://github.com/Microsoft/vscode/pull/59236

Unless I missing something, we can now set the role of the individual items but not the tree itself, although it looks like you can for a pure list.  Whack me if I have missed something.

I can change to use the new getAriaRole callback until we see if anyone objects to me adding the role

**Comment by isidorn** [comment]
After investigation and testing I had to revert the PR. Reason being that the VoiceOver on mac only properly reads out things for the combination of tree / treeitem roles. For list / listitem roles and for checkbox it is silent all the time.
Using checkbox on windows also keeps nvda silent.
Thus it seems it does not make sense to customize these roles since it breaks screen reader behavior
Due to this -> reopening this issue

**Comment by cleidigh** [comment]
@isidorn 
cc: @roblourens 
you ran into several of the things I did when working on the new settings editor which added some different ARIA requirements.  I got things to work for all input types with a few games.  You can see these working on both Windows and OS X.  As I mentioned previously I found that there are plenty of inconsistencies in how the readers deal with input elements that are descendents of widget roles such as tree.  After experimenting with a lot of different combinations I got each of the inputs to work.  In the case of settings I used a primary role of form but I also had to change some of the individual element roles including checkbox.

I am more handicapped on OS X so I was not yet able to make an example there, but I just have to do similar things to what we did with settings.  I think I can get this to work in the same fashion.  I will play with it for October.  

Check out the new settings editor with voiceover...

**Comment by isidorn** [comment]
@cleidigh got it. Feel free to do a PR and ping me on it and we can look into it in October

**Comment by isidorn** [comment]
Moving to november, not something for endgame

**Comment by roblourens** [comment]
The role and 'isChecked' seem to be seen by VoiceOver, but I don't hear it read the set size or posInSet, dunno if that's an issue.
